sage.py

```diff
@@ -571,393 +485,271 @@ def getepisodes(action_seq, mcat, plot, debug=False):
+    print('\n# sub-sequences', len(attacks))
+    return subsequences, attacks
+
+
+## 27 Aug 2020: Generating traces for FlexFringe
+def generate_traces(subsequences, datafile):
+    all_services = [[most_frequent(epi[6]) for epi in subseq] for subseq in subsequences]
+    print('----- All unique services -----')
+    print(set([service for services_subseq in all_services for service in services_subseq]))
     print('---- end ----- ')
 
-    count_lines = 0
-    count_cats = set()
+    num_traces = 0
+    unique_mcat_mserv = set()  # FlexFringe treats the (mcat,mserv) pairs as symbols of the alphabet
 
-    f = open(datafile, 'w')
-    lengths = []
-    lines = []
-    for i,episodes in enumerate(victims):
-        #print(episodes)
-        num_behav = keys[i].split('-')[-1]
-        #print(keys[i], num_behav)
+    episode_traces = []
+    for i, episodes in enumerate(subsequences):
         if len(episodes) < 3:
             continue
-        #lengths+= len(episodes)
-        count_lines += 1
-        mcats = [str(x[2]) for x in episodes]
-        num_servs = [len(set((x[6]))) for x in episodes]
-        max_servs = [most_frequent(x[6]) for x in episodes]
+        num_traces += 1
+        mcats = [x[2] for x in episodes]
+        num_services = [len(set((x[6]))) for x in episodes]
+        max_services = [most_frequent(x[6]) for x in episodes]
         stime = [x[0] for x in episodes]
-        #print(stime)
-        #print(' '.join(mcats))
-
-        #multi = [str(c)+":"+str(n)+","+str(s) for (c,s,n) in zip(mcats,max_servs,num_servs)] # multivariate case
-        multi = [str(small_mapping[int(c)])+"|"+str(s) for (c,s,n,st) in zip(mcats,max_servs,num_servs, stime)] # merging mcat and serv into one
-        #print(multi)
-        for e in multi:
-            feat = e.split(':')[0]
-            count_cats.add(feat)
-        multi.reverse()
-        st = '1' + " "+ str(len(mcats)) + ' ' + ' '.join(multi) + '\n'
-        #f.write(st)
-        lines.append(st)
-    f.write(str(count_lines) + ' ' + str(len(count_cats)) + '\n')
-    for st in lines:
-        f.write(st)
+
+        #multi = [str(c) + ":" + str(n) + "," + str(s) for (c, s, n) in zip(mcats, max_services, num_services)] # multivariate case
+        mcat_mserv_pairs = [small_mapping[mcat] + "|" + mserv for mcat, mserv in zip(mcats, max_services)]
+        unique_mcat_mserv.update(mcat_mserv_pairs)
+        mcat_mserv_pairs.reverse()  # Reverse traces to accentuate high-severity episodes (to create an S-PDFA)
+        trace = '1' + " " + str(len(mcats)) + ' ' + ' '.join(mcat_mserv_pairs) + '\n'
+        episode_traces.append(trace)
+
+    f = open(datafile, 'w')
+    f.write(str(num_traces) + ' ' + str(len(unique_mcat_mserv)) + '\n')
+    for trace in episode_traces:
+        f.write(trace)
     f.close()
-    #print(lengths, lengths/float(count_lines))
-    
+
+
 ## 2 sept 2020: Learning the model
 def flexfringe(*args, **kwargs):
-  """Wrapper to call the flexfringe binary
```

Comment:
This method was not indented with 4 whitespaces (this diff might be a bit confusing)